File: nowcode/huazhong0415/L.py
#pragma clang diagnostic push
#pragma ide diagnostic ignored "cppcoreguidelines-narrowing-conversions"
#pragma ide diagnostic ignored "hicpp-signed-bitwise"
#pragma GCC optimize ("Ofast,unroll-loops")
#pragma GCC optimize("no-stack-protector,fast-math")
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<ll, ll> pll;
typedef pair<int, int> pii;
typedef pair<double, double> pdd;
typedef vector<int> vi;
typedef vector<ll> vll;
typedef vector<double> vd;
typedef vector<vi> vvi;
typedef vector<vll> vvll;
typedef vector<pii> vpii;
typedef vector<pll> vpll;
typedef vector<pdd> vpdd;
typedef vector<vd> vvd;
#define yn(ans) printf("%s\n", (ans)?"Yes":"No");
#define YN(ans) printf("%s\n", (ans)?"YES":"NO");
template<class T> bool chmax(T &a, T b) {
    if (a >= b) return false;
    a = b; return true;
}
template<class T> bool chmin(T &a, T b) {
    if (a <= b) return false;
    a = b; return true;
}
#define FOR(i, s, e, t) for ((i) = (s); (i) < (e); (i) += (t)) 
#define REP(i, e) for (int i = 0; i < (e); ++i) 
#define REP1(i, s, e) for (int i = (s); i < (e); ++i)
#define RREP(i, e) for (int i = (e); i >= 0; --i)
#define RREP1(i, e, s) for (int i = (e); i >= (s); --i)
#define all(v) v.begin(), v.end()
#define pb push_back
#define qb pop_back
#define pf push_front
#define qf pop_front
#define maxe max_element
#define mine min_element
ll inf = 1e18;
#define DEBUG printf("%d\n", __LINE__); fflush(stdout);
template<class T> void print(vector<T> &v, bool withSize = false) {
    if (withSize) cout << v.size() << endl;
    REP(i, v.size()) cout << v[i] << " "; 
    cout << "\n";
}
mt19937_64 rng((unsigned int) chrono::steady_clock::now().time_since_epoch().count());

int __FAST_IO__ = []() {
    std::ios::sync_with_stdio(0);
    std::cin.tie(0);
    std::cout.tie(0);
    return 0;
}();

// Mint & Combinatorics
 
template <typename T>
T inverse(T a, T m) {
  T u = 0, v = 1;
  while (a != 0) {
    T t = m / a;
    m -= t * a; swap(a, m);
    u -= t * v; swap(u, v);
  }
  return u;
}
 
template <typename T>
class Modular {
 public:
  using Type = typename decay<decltype(T::value)>::type;
 
  constexpr Modular() : value() {}
  template <typename U>
  Modular(const U& x) {
    value = normalize(x);
  }
 
  template <typename U>
  static Type normalize(const U& x) {
    Type v;
    if (-mod() <= x && x < mod()) v = static_cast<Type>(x);
    else v = static_cast<Type>(x % mod());
    if (v < 0) v += mod();
    return v;
  }
 
  const Type& operator()() const { return value; }
  template <typename U>
  explicit operator U() const { return static_cast<U>(value); }
  constexpr static Type mod() { return T::value; }
 
  Modular& operator+=(const Modular& other) { if ((value += other.value) >= mod()) value -= mod(); return *this; }
  Modular& operator-=(const Modular& other) { if ((value -= other.value) < 0) value += mod(); return *this; }
  template <typename U> Modular& operator+=(const U& other) { return *this += Modular(other); }
  template <typename U> Modular& operator-=(const U& other) { return *this -= Modular(other); }
  Modular& operator++() { return *this += 1; }
  Modular& operator--() { return *this -= 1; }
  Modular operator++(int) { Modular result(*this); *this += 1; return result; }
  Modular operator--(int) { Modular result(*this); *this -= 1; return result; }
  Modular operator-() const { return Modular(-value); }
 
  template <typename U = T>
  typename enable_if<is_same<typename Modular<U>::Type, int>::value, Modular>::type& operator*=(const Modular& rhs) {
#ifdef _WIN32
    uint64_t x = static_cast<int64_t>(value) * static_cast<int64_t>(rhs.value);
    uint32_t xh = static_cast<uint32_t>(x >> 32), xl = static_cast<uint32_t>(x), d, m;
    asm(
      "divl %4; \n\t"
      : "=a" (d), "=d" (m)
      : "d" (xh), "a" (xl), "r" (mod())
    );
    value = m;
#else
    value = normalize(static_cast<int64_t>(value) * static_cast<int64_t>(rhs.value));
#endif
    return *this;
  }
  template <typename U = T>
  typename enable_if<is_same<typename Modular<U>::Type, long long>::value, Modular>::type& operator*=(const Modular& rhs) {
    long long q = static_cast<long long>(static_cast<long double>(value) * rhs.value / mod());
    value = normalize(value * rhs.value - q * mod());
    return *this;
  }
  template <typename U = T>
  typename enable_if<!is_integral<typename Modular<U>::Type>::value, Modular>::type& operator*=(const Modular& rhs) {
    value = normalize(value * rhs.value);
    return *this;
  }
 
  Modular& operator/=(const Modular& other) { return *this *= Modular(inverse(other.value, mod())); }
 
  friend const Type& abs(const Modular& x) { return x.value; }
 
  template <typename U>
  friend bool operator==(const Modular<U>& lhs, const Modular<U>& rhs);
 
  template <typename U>
  friend bool operator<(const Modular<U>& lhs, const Modular<U>& rhs);
 
  template <typename V, typename U>
  friend V& operator>>(V& stream, Modular<U>& number);
 
 private:
  Type value;
};
 
template <typename T> bool operator==(const Modular<T>& lhs, const Modular<T>& rhs) { return lhs.value == rhs.value; }
template <typename T, typename U> bool operator==(const Modular<T>& lhs, U rhs) { return lhs == Modular<T>(rhs); }
template <typename T, typename U> bool operator==(U lhs, const Modular<T>& rhs) { return Modular<T>(lhs) == rhs; }
 
template <typename T> bool operator!=(const Modular<T>& lhs, const Modular<T>& rhs) { return !(lhs == rhs); }
template <typename T, typename U> bool operator!=(const Modular<T>& lhs, U rhs) { return !(lhs == rhs); }
template <typename T, typename U> bool operator!=(U lhs, const Modular<T>& rhs) { return !(lhs == rhs); }
 
template <typename T> bool operator<(const Modular<T>& lhs, const Modular<T>& rhs) { return lhs.value < rhs.value; }
 
template <typename T> Modular<T> operator+(const Modular<T>& lhs, const Modular<T>& rhs) { return Modular<T>(lhs) += rhs; }
template <typename T, typename U> Modular<T> operator+(const Modular<T>& lhs, U rhs) { return Modular<T>(lhs) += rhs; }
template <typename T, typename U> Modular<T> operator+(U lhs, const Modular<T>& rhs) { return Modular<T>(lhs) += rhs; }
 
template <typename T> Modular<T> operator-(const Modular<T>& lhs, const Modular<T>& rhs) { return Modular<T>(lhs) -= rhs; }
template <typename T, typename U> Modular<T> operator-(const Modular<T>& lhs, U rhs) { return Modular<T>(lhs) -= rhs; }
template <typename T, typename U> Modular<T> operator-(U lhs, const Modular<T>& rhs) { return Modular<T>(lhs) -= rhs; }
 
template <typename T> Modular<T> operator*(const Modular<T>& lhs, const Modular<T>& rhs) { return Modular<T>(lhs) *= rhs; }
template <typename T, typename U> Modular<T> operator*(const Modular<T>& lhs, U rhs) { return Modular<T>(lhs) *= rhs; }
template <typename T, typename U> Modular<T> operator*(U lhs, const Modular<T>& rhs) { return Modular<T>(lhs) *= rhs; }
 
template <typename T> Modular<T> operator/(const Modular<T>& lhs, const Modular<T>& rhs) { return Modular<T>(lhs) /= rhs; }
template <typename T, typename U> Modular<T> operator/(const Modular<T>& lhs, U rhs) { return Modular<T>(lhs) /= rhs; }
template <typename T, typename U> Modular<T> operator/(U lhs, const Modular<T>& rhs) { return Modular<T>(lhs) /= rhs; }
 
template<typename T, typename U>
Modular<T> power(const Modular<T>& a, const U& b) {
  assert(b >= 0);
  Modular<T> x = a, res = 1;
  U p = b;
  while (p > 0) {
    if (p & 1) res *= x;
    x *= x;
    p >>= 1;
  }
  return res;
}
 
template <typename T>
bool IsZero(const Modular<T>& number) {
  return number() == 0;
}
 
template <typename T>
string to_string(const Modular<T>& number) {
  return to_string(number());
}
 
// U == std::ostream? but done this way because of fastoutput
template <typename U, typename T>
U& operator<<(U& stream, const Modular<T>& number) {
  return stream << number();
}
 
// U == std::istream? but done this way because of fastinput
template <typename U, typename T>
U& operator>>(U& stream, Modular<T>& number) {
  typename common_type<typename Modular<T>::Type, long long>::type x;
  stream >> x;
  number.value = Modular<T>::normalize(x);
  return stream;
}
 
struct MOD {
    static int value;
};
int MOD::value = 1e9+7;
using Mint = Modular<MOD>;
typedef vector<Mint> vm;
typedef vector<vm> vvm;

vector<Mint> fac, ifac;
void initFac(int N) {
    fac.resize(N + 1);
    ifac.resize(N + 1);
    fac[0] = ifac[0] = 1;
    REP1(i, 1, N + 1) {
        fac[i] = fac[i - 1] * i;
    }
    ifac[N] = 1 / fac[N];
    RREP1(i, N - 1, 1) {
        ifac[i] = (i + 1) * ifac[i + 1];
    }
}
 
Mint C(int m, int n) {
    if (m < n || n < 0) return 0;
    return fac[m] * ifac[n] * ifac[m - n]; 
}

Mint P(int m, int n) {
    if (m < n || n < 0) return 0;
    return fac[m] * ifac[m - n];
}

struct HLD {
    vi p, dep, sz, hson, dfn, top, last, rnk;

    HLD(vector<vvi> &g, int root = 0) {
        int N = g.size();
        p.resize(N, -1);
        dep.resize(N, 0);
        sz.resize(N, 1);
        hson.resize(N, -1);
        
        function<void(int, int)> dfs = [&](int cur, int par) {
            p[cur] = par;
            if (par != -1) dep[cur] = dep[par] + 1;
            for (auto &next: g[cur]) {
                if (next[0] == par) continue;
                dfs(next[0], cur);
                sz[cur] += sz[next[0]];
                if (hson[cur] == -1 || sz[hson[cur]] < sz[next[0]]) {
                    hson[cur] = next[0];
                }
            }
        };
        dfs(root, -1);
        
        dfn.resize(N, -1);
        top.resize(N, -1);
        last.resize(N, -1);
        rnk.resize(N, -1);
        int tm = 0;
        function<void(int, int)> dfs2 = [&](int cur, int tc) {
            top[cur] = tc;
            dfn[cur] = tm;
            last[tc] = tm;
            rnk[tm] = cur;
            tm++;
            if (hson[cur] == -1) return;
            dfs2(hson[cur], tc);
            for (auto &son: g[cur]) {
                int next = son[0];
                if (next != hson[cur] && next != p[cur]) dfs2(next, next);
            }
        };
        dfs2(root, root);
    }
};

// S: return type, op: how to merge at return, e: S default constructor
// F: value type of lazy value
// mapping: called on apply() to each item, mapping(F f, S item)
// composition: called on apply() to lazy value of a seg, composition(F f, F lazy)
// id: default falue of lazy value
template <class S,
          S (*op)(S, S),
          S (*e)(),
          class F,
          S (*mapping)(F, S),
          F (*composition)(F, F),
          F (*id)()>
struct lazy_segtree {
  public:
    lazy_segtree() : lazy_segtree(0) {}
    explicit lazy_segtree(int n) : lazy_segtree(std::vector<S>(n, e())) {}
    explicit lazy_segtree(const std::vector<S>& v) : _n(int(v.size())) {
        log = 0;
        while ((1 << log) < _n) ++log;
        size = 1 << log;
        d = std::vector<S>(2 * size, e());
        lz = std::vector<F>(size, id());
        for (int i = 0; i < _n; i++) d[size + i] = v[i];
        for (int i = size - 1; i >= 1; i--) {
            update(i);
        }
    }

    void set(int p, S x) {
        assert(0 <= p && p < _n);
        p += size;
        for (int i = log; i >= 1; i--) push(p >> i);
        d[p] = x;
        for (int i = 1; i <= log; i++) update(p >> i);
    }

    S get(int p) {
        assert(0 <= p && p < _n);
        p += size;
        for (int i = log; i >= 1; i--) push(p >> i);
        return d[p];
    }

    S prod(int l, int r) {
        assert(0 <= l && l <= r && r <= _n);
        if (l == r) return e();

        l += size;
        r += size;

        for (int i = log; i >= 1; i--) {
            if (((l >> i) << i) != l) push(l >> i);
            if (((r >> i) << i) != r) push((r - 1) >> i);
        }

        S sml = e(), smr = e();
        while (l < r) {
            if (l & 1) sml = op(sml, d[l++]);
            if (r & 1) smr = op(d[--r], smr);
            l >>= 1;
            r >>= 1;
        }

        return op(sml, smr);
    }

    S all_prod() { return d[1]; }

    void apply_single(int p, F f) {
        assert(0 <= p && p < _n);
        p += size;
        for (int i = log; i >= 1; i--) push(p >> i);
        d[p] = mapping(f, d[p]);
        for (int i = 1; i <= log; i++) update(p >> i);
    }
    void apply(int l, int r, F f) {
        assert(0 <= l && l <= r && r <= _n);
        if (l == r) return;

        l += size;
        r += size;

        for (int i = log; i >= 1; i--) {
            if (((l >> i) << i) != l) push(l >> i);
            if (((r >> i) << i) != r) push((r - 1) >> i);
        }

        {
            int l2 = l, r2 = r;
            while (l < r) {
                if (l & 1) all_apply(l++, f);
                if (r & 1) all_apply(--r, f);
                l >>= 1;
                r >>= 1;
            }
            l = l2;
            r = r2;
        }

        for (int i = 1; i <= log; i++) {
            if (((l >> i) << i) != l) update(l >> i);
            if (((r >> i) << i) != r) update((r - 1) >> i);
        }
    }

    template <bool (*g)(S)> int max_right(int l) {
        return max_right(l, [](S x) { return g(x); });
    }
    template <class G> int max_right(int l, G g) {
        assert(0 <= l && l <= _n);
        assert(g(e()));
        if (l == _n) return _n;
        l += size;
        for (int i = log; i >= 1; i--) push(l >> i);
        S sm = e();
        do {
            while (l % 2 == 0) l >>= 1;
            if (!g(op(sm, d[l]))) {
                while (l < size) {
                    push(l);
                    l = (2 * l);
                    if (g(op(sm, d[l]))) {
                        sm = op(sm, d[l]);
                        l++;
                    }
                }
                return l - size;
            }
            sm = op(sm, d[l]);
            l++;
        } while ((l & -l) != l);
        return _n;
    }

    template <bool (*g)(S)> int min_left(int r) {
        return min_left(r, [](S x) { return g(x); });
    }
    template <class G> int min_left(int r, G g) {
        assert(0 <= r && r <= _n);
        assert(g(e()));
        if (r == 0) return 0;
        r += size;
        for (int i = log; i >= 1; i--) push((r - 1) >> i);
        S sm = e();
        do {
            r--;
            while (r > 1 && (r % 2)) r >>= 1;
            if (!g(op(d[r], sm))) {
                while (r < size) {
                    push(r);
                    r = (2 * r + 1);
                    if (g(op(d[r], sm))) {
                        sm = op(d[r], sm);
                        r--;
                    }
                }
                return r + 1 - size;
            }
            sm = op(d[r], sm);
        } while ((r & -r) != r);
        return 0;
    }

  private:
    int _n, size, log;
    std::vector<S> d;
    std::vector<F> lz;

    void update(int k) { d[k] = op(d[2 * k], d[2 * k + 1]); }
    void all_apply(int k, F f) {
        d[k] = mapping(f, d[k]);
        if (k < size) lz[k] = composition(f, lz[k]);
    }
    void push(int k) {
        all_apply(2 * k, lz[k]);
        all_apply(2 * k + 1, lz[k]);
        lz[k] = id();
    }
};

vector<vector<vector<int>>> readGraphWithWeight(int N, int M, bool isDirected = false) {
    vector<vvi> g(N);
    REP(i, M) {
        int u, v, w;
        cin >> u >> v >> w;
        --u, --v;
        g[u].push_back({v, w});
        if (!isDirected) g[v].push_back({u, w});
    }
    return g;
}

struct Node {
    Mint val, sum;
    int len;
    Node (): val(0), sum(0), len(0) {}
    Node (int v): val(v * v), sum(v), len(1) {}
};

Node op(Node a, Node b) {
    if (a.len == 0) return b; else if (b.len == 0) return a;
    Node c;
    c.len = a.len + b.len;
    c.val = a.val + b.val;
    c.sum = a.sum + b.sum;
    return c;
}

Node e() {return Node();}

Node mapping(pii a, Node b) {
    if (a.second == 2) {
        Node c;
        c.len = b.len;
        c.sum = b.sum + Mint(1) * a.first * b.len;
        c.val = b.val + Mint(2) * a.first * b.sum + Mint(b.len) * a.first * a.first; 
        return c;
    } else if (a.second == 3) {
        Node c;
        c.len = b.len;
        c.sum = Mint(1) * b.len * a.first;
        c.val = Mint(1) * b.len * a.first * a.first;
        return c;
    } else {
        return b;
    }
}

pii comp(pii a, pii b) {
    if (a.second == 0) return b;
    else if (b.second == 0) return a;
    if (a.second == 3) return a;
    else return {a.first + b.first, b.second};
} 

pii id() {return {0, 0};}

int main() {
    int N, M;
    cin >> N >> M;
    auto g = readGraphWithWeight(N, N - 1);
    
    HLD h(g);
    lazy_segtree<Node, op, e, pii, mapping, comp, id> seg(N);
    REP(i, N) {
        for (auto &e: g[i]) {
            if (h.dep[e[0]] == h.dep[i] - 1) seg.set(h.dfn[i], Node(e[1]));
        }
    }
    
    REP(i, M) {
        int op, x, y, z;
        cin >> op >> x >> y;
        --x, --y;
        if (op != 1) cin >> z;
        Mint ans = 0;
        while (h.top[x] != h.top[y]) {
            if (h.dep[h.top[x]] > h.dep[h.top[y]]) {
                if (op == 1) {
                    Node tmp = seg.prod(h.dfn[h.top[x]], h.dfn[x] + 1);
                    ans += tmp.val;
                } else {
                    seg.apply(h.dfn[h.top[x]], h.dfn[x] + 1, {z, op});
                }
                x = h.p[h.top[x]];  
            } else {
                if (op == 1) {
                    Node tmp = seg.prod(h.dfn[h.top[y]], h.dfn[y] + 1);
                    ans += tmp.val;
                } else {
                    seg.apply(h.dfn[h.top[y]], h.dfn[y] + 1, {z, op});
                }
                y = h.p[h.top[y]];
            }
        }
        if (h.dep[x] < h.dep[y]) swap(x, y);
        if (op == 1) {
            Node tmp = seg.prod(h.dfn[y] + 1, h.dfn[x] + 1);
            ans += tmp.val;
        } else {
            seg.apply(h.dfn[y] + 1, h.dfn[x] + 1, {z, op});
        }
        if (op == 1) printf("%d\n", ans);
    }
    return 0;
}

File: nowcode/shandong/l.cpp
#pragma GCC optimize("Ofast,unroll-loops")
#include <bits/stdc++.h>
using namespace std;

using ll = long long;
#define ar(x) array<int,x> 
#define all(c) (c).begin(), (c).end()
#define rall(x) (x).rbegin(), (x).rend() 
#define sz(x) (int)(x).size()
#define f0(e) for(int i = 0; i < (e); ++i)
#define f1(e) for(int i = 1; i <= (e); ++i)
#define f2(i,e) for(int i = 0; i < (e); ++i)
#define f3(i,a,b) for (int i=(a);i<(b);i++)
#define r3(i,b,a) for (int i=(b)-1;i>=(a);i--)
#define Sm(a) accumulate((a).begin(), (a).end() , 0ll)
#define Mn(a) (*min_element((a).begin(), (a).end()))
#define Mx(a) (*max_element((a).begin(), (a).end()))
#define rev(a) reverse((a).begin(), (a).end())
#define each(x,a) for(auto& x : a)
mt19937 mrand(random_device{}()); 
int rng(int x) { return mrand() % x;}
int pct(long long x) {return __builtin_popcountll(x);} 
int lg(int x) {return x == 0 ? -1 : 31 - __builtin_clz(x);}
int clg(int x) {return x <= 1 ? 0 : 32 - __builtin_clz(x - 1);}
template<class t,class u> bool cmx(t &a, const u &b){return a < b ? a = b, 1 : 0;}
template<class t,class u> bool cmn(t &a, const u &b){return b < a ? a = b, 1 : 0;}
template <class T> int lb(const vector<T> &v, const T &x) { return distance(begin(v), lower_bound(begin(v), end(v), x));}
template <class T> int rb(const vector<T> &v, const T &x) { return distance(begin(v), upper_bound(begin(v), end(v), x));}
template<class T,class A> void psum(vector<T>& s, const vector<A>&a){
    int n=a.size();s.resize(n+1);s[0]=0;for(int i=0;i<n;++i)s[i+1]=s[i]+a[i];  
};
template<typename T, typename F> T b_search(T l, T r, bool fst, F &&f) {
    T c = fst ? r : l;
    while(l<=r){T md=(l+r)/2;if(f(md)){c=md; fst?(r=md-1):(l=md+1);} else fst?(l=md+1):(r=md-1);}
    return c;
}
template <typename T, typename U> // pair
ostream& operator<<(ostream& out, const pair<T, U>& a) {return out << a.first << ' ' << a.second;}
template <typename T>  // vector
ostream& operator<<(ostream& out, const vector<T>& a) {for(int i=0,n=a.size(); i<n; ++i) out<<a[i]<<" \n"[i==n-1]; return out;}
template <typename T, size_t N> // array
ostream& operator<<(ostream& out, const array<T, N>& a) {for(int i=0;i<N;++i)cout<<a[i]<<" \n"[i==N-1];return out;}
template <typename T, typename U> // pair
istream& operator>>(istream& in, pair<T, U>& a) {return in >> a.first >> a.second;}
template <typename T> // vector
istream& operator>>(istream& in, vector<T>& a) {for (auto &x: a) in >> x; return in;}
template <typename T, size_t N> // array
istream& operator>>(istream& in, array<T, N>& a) {for (int i=0;i<N;++i) in >> a[i]; return in;}
template<class T> void rd(T& x) {cin >> x;}
template<class H, class... T> void rd(H& h, T&... t) {rd(h); rd(t...);}
template<class A> void wt(A x) { cout << x;}
template<class H, class... T> void wt(const H& h, const T&... t) { wt(h); wt(t...);}
#ifdef sigma-yyf
#define dbg(...) debug_impl(#__VA_ARGS__, __VA_ARGS__)
template <class H, class... Ts> void debug_impl(const char* s, const H& h, const Ts&... t) {
    cerr << s << ": " << h, ((cerr << ", " << t), ..., (cerr << "\n"));
}
#else
#define dbg(...) void(0)
#endif
struct fast_ios { 
    fast_ios(){ cin.tie(nullptr), ios::sync_with_stdio(false), cout << fixed << setprecision(10);};
} fast_ios_;
constexpr char nl = '\n';
constexpr int inf = 1'061'109'567;
constexpr ll infll = 2'000'000'000'000'000'000;
constexpr int N = 2e5 + 5;
int n, m, q;
int x, y, k;
string s, t;

long long arith_seq_sum(long long a1, long long d, long long n) {
    return a1 * n + n * (n - 1) / 2 * d;
}
#define int long long
void spiralOrder(vector<vector<int>>& mx) {
    int m = mx.size(), n = m ? mx[0].size() : 0;
    int u = 0, d = m - 1, l = 0, r = n - 1, p = 0;
    while (u <= d && l <= r) {
        for (int i = l; i <= r; ++i) mx[u][i] = ++p;
        if (++u > d) break;
        for (int i = u; i <= d; ++i) mx[i][r] = ++p;
        if (--r < l) break;
        for (int i = r; i >= l; --i) mx[d][i]=++p;
        if (--d < u) break;
        for (int i = d; i >= u; --i) mx[i][l]=++p;
        ++l;
    }
}

void ac_yyf(int tt){
    ll n,m,x;
    rd(n,m,x);
    if(n*1ll*m<=100) {
        vector<vector<int>> f(n,vector<int>(m));
        spiralOrder(f);
        f0(n)f2(j,m)if(f[i][j]==x){
            wt(i+1,' ',j+1,nl);
            return;
        }
    }
    int t=min(n,m);

    if(n==1) {
        wt("1 ",x,nl);
        return;
    }

    if(m==1) {
        wt(x, " 1",nl);
        return;
    }

    int g=(t+1)/2;
    ll a1=n+n+m+m-4;
    ll a2=n-2+n-2+m-2+m-2-4;
    ll d=a2-a1;

    if(x<=a1){
        if(x<=m){
            wt("1 ",x,nl);
        }else if(x<=m+n-1){
            wt(x-m+1,' ',m,nl);
        }else if(x<=m+n-1+m-1){
            ll d=x-(m+n-1);
            wt(n,' ',m-d,nl);
        }else{
            ll d=x-(m+n-1+m-1);
            // dbg(d,n,m);
            wt(n-d,' ',1,nl);
        }
        return;
    }
    ll s=0;
    auto c=b_search<ll>(1,g,0,[&](ll t1){
        ll s1=arith_seq_sum(a1,d,t1);
        return s1<x;
    });
    s=arith_seq_sum(a1,d,c);

    // if(x==n*1ll*m&&(t&1)){
    //     wt(c+1,' ',c+1,nl);
    //     return;
    // }

    // x在第c+1圈。

    ll left=x-s;

    // 第c+1圈的第一个数字为 s+1
    // 第c+1圈的长宽分别为n1,m1

    int n1=n-2*c;
    int m1=m-2*c;

    if(left<=m1) {
        wt(c+1,' ',left+c,nl);
    }else if(left<=m1+n1-1) {
        ll d=left-m1;
        wt(c+1+d,' ',m-c,nl);
    }else if(left<=m1+n1-1+m1-1){
        ll d=left-(m1+n1-1);
        wt(n-c,' ',m-c-d,nl);
    }else{
        ll d=left-(m1+n1-1+m1-1);
        wt(n-c-d,' ',c+1,nl);
    }

    // dbg(a1,a2,nl);
}
signed main() {
    int T = 1, fun_type = 0;
    cin >> T;
    for (int cas = 1; cas <= T; ++cas) {
        ac_yyf(cas);
        // cout << (ac_yyf(cas) ? "Alice" : "Bob") << '\n';
        // cout << (ac_yyf(cas) ? "YES" : "NO") << '\n';
    }
    return 0;
}
